From c56b1e6d2b1803de086143ba3f7b806bc0b9036b Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Wed, 3 Feb 2016 16:09:57 +0800
Subject: [PATCH 0942/5242] MLK-12344-3 usb: chipidea: update power lost
 handling for gadget

commit  405d8dba798a05e229fc3cb2d711d16c18a2a688 from
https://source.codeaurora.org/external/imx/linux-imx.git

After we put gadget disconnect and connect in id switch handling,
update power lost work accordingly.

Acked-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
(cherry picked from commit fd49596eceeb4e61267f35fb74bf0c04056f2ed7)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/core.c |    8 +++++---
 drivers/usb/chipidea/udc.c  |    4 ----
 2 files changed, 5 insertions(+), 7 deletions(-)

diff --git a/drivers/usb/chipidea/core.c b/drivers/usb/chipidea/core.c
index e284e94..3f664c2 100644
--- a/drivers/usb/chipidea/core.c
+++ b/drivers/usb/chipidea/core.c
@@ -940,11 +940,13 @@ static void ci_start_new_role(struct ci_hdrc *ci)
 {
 	enum ci_role role = ci_get_role(ci);
 
-	if (ci->role != role)
+	if (ci->role != role) {
 		ci_handle_id_switch(ci);
-
-	if (role == CI_ROLE_GADGET)
+	} else if (role == CI_ROLE_GADGET) {
+		if (ci->vbus_active)
+			usb_gadget_vbus_disconnect(&ci->gadget);
 		ci_handle_vbus_connected(ci);
+	}
 }
 
 static void ci_power_lost_work(struct work_struct *work)
diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index f711443..5366e33 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -2105,10 +2105,6 @@ static void udc_suspend_for_power_lost(struct ci_hdrc *ci)
 /* Power lost with device mode */
 static void udc_resume_from_power_lost(struct ci_hdrc *ci)
 {
-	/* Force disconnect if power lost with vbus on */
-	if (!ci_otg_is_fsm_mode(ci) && ci->vbus_active)
-		usb_gadget_vbus_disconnect(&ci->gadget);
-
 	if (ci->is_otg)
 		hw_write_otgsc(ci, OTGSC_BSVIS | OTGSC_BSVIE,
 					OTGSC_BSVIS | OTGSC_BSVIE);
-- 
1.7.9.5

